### ISSUE TITLE
Type inconsistencies in implementations of ValueContainer#getValue?

### ISSUE THREAD

**Issue by kory33** [SpongePowered/SpongeAPI#1824]
Hello, I have a question on Data API.

I wanted to implement my own version of abstract data manipulator. I've already verified that it works for simple data constructs. While writing type proofs, I've found that it is not easy to ensure that the return value of `getValue` is `Optional<V>`. The question is; is it OK to return `Optional<Value<E>>` or `Optional<ImmutableValue<E>>` instead of `Optional<V>` in `ValueContainer#getValue`?

Looking at SpongeAPI code, it does the same thing as I do; it actually returns `Optional<Value<E>>` instead of `Optional<V>`: [AbstractData#getValue](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/data/manipulator/mutable/common/AbstractData.java#L164), and `Optional<ImmutableValue<E>>` instead of `Optional<V>`: [AbstractImmutableData#getValue](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/data/manipulator/immutable/common/AbstractImmutableData.java#L113)

In the former code for example, shouldn't there be a case where `V` is `ImmutableValue<E>` and the code leads to a potential heap pollution? If so, may I propose the signature of `ValueContainer#getValue` to be

```Java
<E> Optional<? extends BaseValue<E>> getValue(Key<? extends BaseValue<E>> key)
```

instead?

**Comment by gabizou** [comment]
It's never OK to return a weaker type. The abstract implementation provided in the API is a lose contracted and *semi functioning* implementation to provide developers ways to implement their own Data without having to write all of the boilerplate code themselves. It's been intended to strip away the abstract implementation in #1198, but it's not completed as other priorities take precedence. 

To answer the question about whether the implementation does cause heap pollution: Short answer is **no**, but the long answer is dependent on the developer using the abstract implementation. If the plugin developer is returning a `Value` of a weaker type than the `Key` supports, that is their fault as the generic contract in their own created `Key` instances. Ultimately, Sponge's implementation of all Sponge provided Data does assure that heap pollution does not take place, because the return types of the `Supplier`s are matching the `Value` types identified by the `Key`.

It would be a good type check to use the `Key`'s `TypeToken` to validate that the supplier is returning the correct type, and point it out to the plugin developer, but in short, the implementation is not responsible for what is ultimately considered "user error".

Proposing the change of signature isn't advisable as it breaks usability of the multiple types of possible `Value`s available from `Key`s and restricts developers to having to do the type checking themselves.

**Comment by kory33** [comment]
Thank you for the reply. Now I see why the signature is like that.

I'm not yet clear with the (currently provided) implementation of Abstract(Immutable)Data.
If I'm not mistaken, `AbstractData#getValue` *always* returns a `Value` wrapped in `Optional`, and `AbstractImmutableData#getValue` *always* returns an `ImmutableValue` wrapped in `Optional`(I recognize that ensuring `E` is upto the plugin developer).

Does this imply that, for example, given a `key` that expects a `Value<Int>` from one `manipulator`, `manipulator.asImmutable().getValue(key)` should equal `Optional.empty()` because the immutable version does not have the corresponding mutable `Value`?

**Comment by gabizou** [comment]
> If I'm not mistaken, AbstractData#getValue always returns a Value wrapped in Optional, and AbstractImmutableData#getValue always returns an ImmutableValue wrapped in Optional(I recognize that ensuring E is upto the plugin developer).


You're confusing the *declared type* versus the *instance type*. Yes the internal `Map` declares the entries as `Supplier<Optional<Value<?>>` etc. but the actual return type has the contract of *at least returning the type `Value`* whereas the actual implementation of many [differs based on the `Key` used as demonstrated in some implementations](https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/data/manipulator/mutable/SpongeFireworkRocketData.java#L51) which end up being type checked based on the contracted methods.




> Does this imply that, for example, given a key that expects a Value<Int> from one manipulator, manipulator.asImmutable().getValue(key) should equal Optional.empty() because the immutable version does not have the corresponding mutable Value?

Remember that the Values and Manipulators are contracted to have guaranteed equivalency when transferring from mutable -> immutable -> mutable such that the values are the same, and if the value is a bounded value, the bounds would likewise be the same. The one aspect of differences is that we're always passing the underlying raw value (whether it's an `int`, `boolean`, `String`, `List<PotionEffect>` etc.) to create either a new `Value`/`ImmutableValue` or [if it's possible to retrieve a cached `ImmutableValue`](https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/data/ImmutableDataCachingUtil.java#L98).

**Comment by kory33** [comment]
hmm, I've tried the following code in `spongevanilla-1.12.2-7.1.0-BETA-22`(it's Kotlin code but I don't think that matters much):
```Kotlin
causePlayer.get(SpongeHealthData::class.java).map {
    val healthKey: Key<MutableBoundedValue<Double>> = Keys.HEALTH
    val mutableManipulatorValue: Optional<MutableBoundedValue<Double>> = it.getValue(healthKey)
    val immutableManipulatorValue: Optional<MutableBoundedValue<Double>> = it.asImmutable().getValue(healthKey)

    println(mutableManipulatorValue)
    println(immutableManipulatorValue)
    val unwrapped: MutableBoundedValue<Double> = immutableManipulatorValue.get()
}
```
Declared types of those `val` variables are all inferred types. As the types suggest, I would expect `MutableBoundedValue` to be in both optionals.
However, what is logged is
```
[STDOUT]: Optional[SpongeBoundedValue{key=SpongeKey{..., valueToken=...mutable.MutableBoundedValue<....Double>, ...]
[STDOUT]: Optional[ImmutableSpongeBoundedValue{key=SpongeKey{..., valueToken=...mutable.MutableBoundedValue<....Double>, ...]
```
Therefore at the last line, `ClassCastException` is thrown. Is this still what is intended?

**Comment by gabizou** [comment]
Yes... You'd still get a `ClassCastException` because you're declaring the type as `ImmutableBoundedValue<Double>`, which is what the [return type would be](https://github.com/SpongePowered/SpongeAPI/blob/7ac34cb45149e3891e79d005e2a36f4ac7bb0d76/src/main/java/org/spongepowered/api/data/value/mutable/MutableBoundedValue.java#L38) (the pointer is to the latest commit that has edited the file, so that if bleeding ever changes in the future, the link will still work).

The other problem is that relying on the generics for the Key to inflect the value retrieved is not going to work in this case because the `Key` declares the generic as a mutable value, even though the value returned from the `ImmutableDataManipulator`/`ImmutableDataHolder` would always return an `ImmutableValue` or subclass thereof. Granted, the java docs are not clear on that aspect, but your local variable declaration is still valid for this line:
```kotlin
    val immutableManipulatorValue: Optional<MutableBoundedValue<Double>> = it.asImmutable().getValue(healthKey)

```
but the result will still be the same as explained above. The return value at runtime will be an immutable variant. Without breaking virtually all of Data API (due to the key values being declared with how they are), I could make the breaking change to where there's a base for the extraneous specific value types without them being mutable or immutable and only available as retrieval without being able to set the value unless you type check for mutability. The biggest change would be that all of `CollectionValue`, `MapValue`, `ListValue`, etc. would be changed for packages, and then likewise, we'd be making `MutableCollectionValue` etc.

**Comment by kory33** [comment]
I see. 

> The other problem is that relying on the generics for the Key to inflect the value retrieved is not going to work in this case because the Key declares the generic as a mutable value, even though the value returned from the ImmutableDataManipulator/ImmutableDataHolder would always return an ImmutableValue or subclass thereof.

So that goes back to my original question; is this implementation ok?

[edit] by "this implementation", I mean the implementation in which the mutable manipulator returns `Optional<Value<E>>` and the immutable manipulator returns `Optional<ImmutableValue<E>>` both casted to the former type when `getValue` is called with `Key<Value<E>>`.

**Comment by kory33** [comment]
Let me rephrase the question...

Suppose I have an implementation of immutable manipulator which returns `ImmutableValue<E>` wrapped in the `Optional` when inquired a value with `getValue(Key<Value<E>>)`. Should such an implementation be avoided and should it *always* return `Value<E>` wrapped in the `Optional`, unlike the code in SpongeCommon does currently?

**Comment by gabizou** [comment]
Re-opening because I may end up breaking Values API by splitting up the different types into immutable base types and then mutable types where `asMutable()` will be available (by default, all values would be immutable in a small regard, unless casting is considered), but we will have strict `Immutable` types still. This way we can provide the guarantee based on the `ValueContainer`s mutability preference (whether it's an immutable or mutable variant) we can appropriately expect the mutability of said values based on the parent. 

This should resolve my other issue with #1198 #1620 or whichever issue I had been thinking about.

**Comment by gabizou** [comment]
Closing as this will be resolved when #1843 is merged.
